### PR TITLE
Match torrents via regular expressions

### LINKED ISSUE #83: Feature request: support regex in search box
Currently the search box at the top-right only supports literal match, while it would be extremely useful to have it supporting regular expressions, so that we could search easily among downloads

thanks!

### PR DESCRIPTION
**Pull request by sandrotosi**
Fix #83

### CODE REVIEW

**Comment by TingPing** [comment]
`GRegex` is deprecated, use PCRE2 directly.

**Comment by sandrotosi** [comment]
no thanks: i've spent already enough time to figure out how to use gregex and where in the trg source code i needed to make this change, please go ahead and make the needed changed to use PCRE2 directly. also note there are other occurrences of GRegex in trg source code, so it's _not_ totally out of line to keep using it

**Comment by TingPing** [comment]
> also note there are other occurrences of GRegex in trg source code, so it's not totally out of line to keep using it

Thats fair.

**Comment by sandrotosi** [comment]
i've actually noticed that if the regex entered is incorret, trg crashed, so we sdhould probably add a sort of try..exept and if in the except branch, just match it as a string and not a regex

File: src/trg-main-window.c [hunk 1]
```diff
@@ -1514,9 +1514,15 @@ trg_torrent_tree_view_visible_func(GtkTreeModel * model,
             gchar *filterCmp = g_utf8_casefold(filterText, -1);
             gchar *nameCmp = g_utf8_casefold(name, -1);
 
-            if (!strstr(nameCmp, filterCmp))
-                visible = FALSE;
+            GRegex *regex;
+            GMatchInfo *match_info;
+            regex = g_regex_new (filterCmp, 0, 0, NULL);
```
Comment (TingPing) [inline src/trg-main-window.c:1519]: Compiling regex every single call is a bit overkill considering the text wouldn't have changed.

File: src/trg-main-window.c [hunk 2]
```diff
@@ -1514,9 +1514,15 @@ trg_torrent_tree_view_visible_func(GtkTreeModel * model,
...
 
+            if (!g_regex_match (regex, nameCmp, 0, &match_info))
```
Comment (TingPing) [inline src/trg-main-window.c:1521]: Just check if `regex` is `NULL` before this call to know if it was valid.

File: src/trg-main-window.c [hunk 3]
```diff
@@ -1514,9 +1514,15 @@ trg_torrent_tree_view_visible_func(GtkTreeModel * model,
...
+	        visible = FALSE;
```
Comment (TingPing) [inline src/trg-main-window.c:1522]: Mixed tabs with spaces.

Comment (TingPing) [inline src/trg-main-window.c:1522]: This logic also assumes its always regex which just isn't true and a string that is valid regex may also be a literal somebody wanted to search for.

I think this needs some UI work honestly to just have the user toggle regex or not.

**Review by TingPing — CHANGES_REQUESTED** [review]

File: src/trg-main-window.c (hunk 1 above)
Comment (sandrotosi) [reply src/trg-main-window.c:1519]: where should the regex be compiled?

File: src/trg-main-window.c (hunk 3 above)
Comment (sandrotosi) [reply src/trg-main-window.c:1522]: it's easy enough to escape regex sequences; the only "common" character usually searched for that's also a special char in regex is `.` , which you can just escape with `\.` (sure also brackets and parentesis could be there, but i would highly unlikely someone would actually search for them: you likely start typing a string not weird special characters).

I'm afraid i will not be able to contribute any UI change.

Comment (TingPing) [reply src/trg-main-window.c:1522]: Regex has plenty of special characters... `()[]+.\`, etc. Its silly to expect users to understand that.

Anyway I'll just close this PR if you won't be doing more in-depth work.

**Comment by sandrotosi** [comment]
after 4 months of silence you come up with extra requirements, extremely more convoluted than the simple approach proposed here (for which we have no proof it would confuse users), provide no guidance or help and then just close the PR -- nice perspective from a first time contributor

i'll just keep my locally recompiled trg which i've been using with profit for long

**Comment by TingPing** [comment]
I do apologize for my unresponsive nature, my life has been hectic and I only volunteer some of it to this project.

That said I do not believe any of my requirements are unreasonable and as you yourself said this patch doesn't even function properly and will crash when given invalid regex.

**Pull request closed without merging**